### DISCUSSION TITLE
Methods on strings?

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#1767] [Ideas]
Imagine if strings were objects with methods on them, so you could do things like:

```python
"hello".upper()
```
Or:
```rust
let hello = "hello"
let helloUpper = hello.upper()
```

**Comment by max-sixty** [comment]
I agree some string functions would be very useful. A survey of what's common enough across DBs to add to the stdlib would be welcome.

For example, maybe we add some functions to the stdlib like `"foo" | upper` which compiles to SQL `UPPER("foo")`

---

At the moment, we only have functions, no methods. That could change in the future, but would be a bigger discussion.

**Reply by vanillajonathan** [reply]
Here is a link to SQL-92 - https://www.contrib.andrew.cmu.edu/~shadow/sql/sql1992.txt
It seems `upper()` and `lower()` have been part of the standard since at least then.

It is supported by MariaDB ([ref](https://mariadb.com/kb/en/lower/)), MySQL ([ref](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html)), Microsoft SQL Server ([ref](https://learn.microsoft.com/en-us/sql/t-sql/functions/lower-transact-sql?view=sql-server-ver16)), PostgreSQL ([ref](https://www.postgresql.org/docs/15/features-sql-standard.html)), and SQLite ([ref](https://sqlite.org/lang_corefunc.html)).

**Reply by max-sixty** [reply]
Great! We could def add those to the stdlib. If you want to open an issue specific to those, that would be welcome. 

(You might be able to add it to the stdlib already, but fine to just open an issue)

**Reply by vanillajonathan** [reply]
Okay I created #1793, but I think it would be nice to have these as methods on the string object, because so many people are used to object orientation, and it would ease discoverability since a IDE could present all the methods that are contextually relevant for the given datatype.

**Reply by max-sixty** [reply]
Re methods — that's a bigger discussion, but briefly:

`"foo" |` can still autocomplete to all functions which have a final parameter of type string. 

So Method vs Function is somewhat fuzzier. The main differences I can see:
- the syntax — `.` vs `|`
- and whether there are some canonical Methods, rather than autocompleting on every compatible function
